### PR TITLE
Migrate off of gogo/protobuf

### PR DESCRIPTION
**Pull request by kzys**
gogo/protobuf has been deprecated since 2022. This change uses Google's code generator instead of gogo/protobuf.

### CODE REVIEW

File: clone.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+func cloneProto[T protoreflect.ProtoMessage](from T) (T, error) {
+	var zero T
+	clone := proto.Clone(from)
+	typed, ok := clone.(T)
+	if !ok {
+		return zero, fmt.Errorf("failed to cast %+v to %+v", clone, zero)
```
Comment (tonistiigi) [inline clone.go:15]: only `pkg/errors` allowed in these repos

File: clone.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import (
+	"fmt"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+func cloneProto[T protoreflect.ProtoMessage](from T) (T, error) {
+	var zero T
+	clone := proto.Clone(from)
```
Comment (tonistiigi) [inline clone.go:12]: I'm a bit worried about including this reflect-based clone in this internal logic where no reflect should be needed. Are there other options?

Comment (tonistiigi) [reply clone.go:12]: Eg. maybe some of these private functions should not use proto based types.

File: clone.go (hunk 1 above)
Comment (kzys) [reply clone.go:15]: I didn't know that. Will fix.

File: clone.go (hunk 2 above)
Comment (kzys) [reply clone.go]: Ended up just doing the below.

```
	clone := &types.Stat{
		Path:     from.Path,
		Mode:     from.Mode,
		Uid:      from.Uid,
		...
	}
```

**Comment by thaJeztah** [comment]
I triggered CI; looks like there's a small linting failure;

```
17.56 receive_test.go:389:6: func `clonePacket` is unused (unused)
17.56 func clonePacket(from *types.Packet) *types.Packet {
17.56      ^
```

**Comment by kzys** [comment]
@thaJeztah Thanks! Fixed.

**Comment by thaJeztah** [comment]
Awesome; I kicked it again, and it's green now 👍 

(I'm not super-familiar with this code, so I'll leave actual review to @tonistiigi, but noticed the PR, and saw CI needed approval 😅)

**Review by tonistiigi — APPROVED** [review]
Do we need some benchmarks as well to ensure there isn't a big regression in marshal/unmarshal? I think in BuildKit I would like some tests that verify backwards compatibility as well (by including binary proto results from previous version).

**Pull request merged**